### PR TITLE
[feat] VSA-256 fastpath on Blackwell via FA4 CuTe block-sparse attention

### PR DESCRIPTION
**Pull request by alexzms**
## Summary

Adds a 256-element VSA tile path that dispatches to the FA4 CuTe block-sparse
attention forward kernel on Blackwell, on top of the existing 64-element
TK/Triton path which remains the default and is unchanged.

Measured ~1.4 PFLOPs forward kernel on a single GB200 (bf16, head_dim=128,
seq_len=40960, 8 heads) for the sparse kernel itself; the 64-tile path is
byte-identical to current behavior at the default settings.

## Design

VSA tile shape flows through the same channel that already carries
`VSA_sparsity`: `FastVideoArgs.VSA_tile_size` (CLI: `--VSA-tile-size T H W`)
→ `attn_metadata_builder.build()` → `VideoSparseAttentionMetadata.VSA_tile_size`
→ backend `forward()`.

- Default `(4, 4, 4)` keeps the existing 64-tile path untouched.
- Setting `(4, 8, 8)` (256-element blocks) triggers the new CuTe BSHD
  fastpath inside `forward()` automatically (dispatch by
  `block_elements = prod(tile_size)`).

Kernel-level backend overrides for debugging / perf comparison (all unset by
default, no behavior change):

- `FASTVIDEO_VSA_TRITON=1` forces Triton (route-A 256→64 expansion for the
  256 path).
- `FASTVIDEO_VSA_TK=1` prefers the sm_90 TK kernel (64 path only).
- `FASTVIDEO_VSA_CUTEDSL=1` prefers CuTe (256 path default).
- `FASTVIDEO_KERNEL_VSA_FORCE_TRITON=1` kept as a backward-compat alias for
  `FASTVIDEO_VSA_TRITON`.

## New dependencies (256-tile path only)

- Submodule `fastvideo-kernel/include/flash-attention` →
  `FoundationResearch/flash-attention` @ `vsapad`. Python-only (CuTe DSL
  JIT); **no C++ build step**.
- Runtime: `nvidia-cutlass-dsl>=4.3` and `quack-kernels`.

The default 64-tile path adds no new dependencies and is unaffected if the
submodule / CuTe DSL is absent.

## Test plan

- [x] `pytest fastvideo-kernel/tests/test_vsa256_forward.py` — CuTe forward vs dense torch reference (2/2)
- [x] `pytest fastvideo-kernel/tests/test_vsa256_forward_vbs.py` — CuTe variable KV block size (1/1)
- [x] `pytest fastvideo-kernel/tests/test_vsa256_triton.py` — Triton fwd+bwd vs torch reference (1/1)
- [x] `pytest fastvideo-kernel/tests/test_vsa256_forward_cross.py` — three-way torch/CuTe/Triton parity (1/1)
- [x] Verified the `(4,4,4)` default path is unchanged vs `main`.
- [ ] SSIM regression suite on Hopper (recommended pre-merge; the 64-tile
      path is untouched but worth confirming).

## Hardware tested

NVIDIA GB200 (sm_100), CUDA 12.9, torch 2.9.1+cu128, aarch64 Linux.

Draft: opening for early review of the dispatch design and the submodule /
dependency story before backward (autograd) support for the CuTe path lands.

### CODE REVIEW

File: .gitmodules
Comment (Davids048) [inline .gitmodules:1]: Could you check if this fork is needed?

Comment (alexzms) [reply .gitmodules:1]: I think I will need to merge the kernel level change of FA4 into our trusted repo https://github.com/hao-ai-lab/flash-attention-fp4

Comment (alexzms) [reply .gitmodules:1]: @Davids048 Good catch — confirmed the fork is **not** needed, and it's been removed.

I traced it through: the submodule pointed at our fork's `vsapad` branch, which turned out to be `Dao-AILab/flash-attention` @ `c19cd20e` plus a single line adding `torchvision` to `cute/pyproject.toml` — i.e. **zero kernel changes**. The `flash_attn.cute` we build against (`interface.py`, `block_sparsity.py`) is byte-identical to that upstream revision, and the VSA-256 parity tests pass against it. (The only real kernel work in the fork is an unrelated `dev/blocksize64` experiment that this PR doesn't use — it ships the 256→128 expansion path, i.e. native block=128.)

So this supersedes my earlier note about merging into `flash-attention-fp4` — there's nothing to merge. The FA4 CuTe backend is now an **optional, lazily-imported** dependency (VSA falls back to Triton when it's absent), documented in `fastvideo-kernel/README.md` (e9320ebf) with a pinned upstream install:

```
pip install "git+https://github.com/Dao-AILab/flash-attention.git@c19cd20e#subdirectory=flash_attn/cute"
```

Pinned to `c19cd20e` because that revision's `_flash_attn_fwd` takes `m_block_size`/`n_block_size` (what the wrapper calls); later upstream reshaped it into a `tile_mn` tuple.

**Comment by SolitaryThinker** [comment]
**Hi @alexzms** — automated review from Gob, one of @SolitaryThinker's AI reviewers. Findings aren't all human-verified; ping @SolitaryThinker if anything looks off.

## TL;DR

Substantial, well-scoped feature PR: clean `block_elements == 256` dispatch, the 64-element TK/Triton path is structurally untouched, the FA4 CuTe dep is correctly lazy + optional, and @Davids048's earlier fork-submodule concern is fully resolved in commits `9560e648e8` + `e9320ebf2b`. Two real things to chase before merge: (1) the CuTe-256 forward has no autograd registration / `requires_grad` guard while the PR carries `scope: training`, and (2) the PR body is stale relative to the post-submodule-removal reality. Both are small surfaces.

## Verdict: **approve-with-followup**

- S0: 0  S1: 0  S2: 3 (3 surfaced)  S3: not shown; see review.md

## Findings

### [S2-important] CuTe-256 forward has no autograd; training silently broken if 256-tile + CuTe is enabled

**What.** `block_sparse_attn_cute_fwd._cute_forward` calls `_flash_attn_fwd` directly without `torch.library.custom_op` + `register_autograd` (the pattern both `block_sparse_attn_triton` and `block_sparse_attn_sm90` follow in `block_sparse_attn.py`). There's no `requires_grad` guard in the kernel and no `self.training` check in `fastvideo/attention/backends/video_sparse_attn.py:forward()`.

**Why it matters.** PR carries `scope: training` label. If a user sets `VSA_TILE_SIZE=(4,8,8)` and `FASTVIDEO_VSA_CUTEDSL=1` together (both opt-in, but both can be set by anyone), forward succeeds but `.backward()` will either raise ("element 0 of tensors does not require grad and does not have a grad_fn") or silently drop sparse-branch gradients — and the BSHD path does `out_view.add_(out_c_blk.unsqueeze(2) * gate_view)` in-place on the CuTe output (`ops.py:226`), which makes the failure mode harder to debug. The Triton-256 fallback (`_triton_via_route_a` → `block_sparse_attn_triton`) IS autograd-safe, so the structural problem is CuTe-only.

**Suggested fix** (pick one):
1. Easiest: add `if q.requires_grad or k.requires_grad or v.requires_grad: raise NotImplementedError("VSA-256 CuTe fastpath is forward-only; unset FASTVIDEO_VSA_CUTEDSL for the Triton route-A autograd path.")` at the top of `_cute_forward`.
2. Better: wrap `_cute_forward` in `torch.library.custom_op` and `register_autograd` raising `NotImplementedError` from the backward — the error then surfaces at `.backward()` with a stack trace pointing here.
3. Drop the `scope: training` label from this PR if CuTe-training is genuinely deferred.

**Evidence:** `fastvideo-kernel/python/fastvideo_kernel/block_sparse_attn_cute_fwd.py:138-184` (no `custom_op`); contrast `fastvideo-kernel/python/fastvideo_kernel/block_sparse_attn.py:104-131,230` (Triton path's autograd registration); `fastvideo/attention/backends/video_sparse_attn.py:271-279` (no `requires_grad` check on the dispatch).

### [S2-important] `kv_blk[0]` indexing in CuTe `_vbs_mask_mod` — please resolve gemini-bot's high-priority flag explicitly

**What.** gemini-bot flagged on `block_sparse_attn_cute_fwd.py:139` (commit `e9320ebf2b`) that `kv_blk` is already a scalar SSA value from `n_idx // block_size_ssa`, and indexing it with `[0]` is likely incorrect. The line: `valid = utils.scalar_to_ssa(kv_sizes[kv_blk[0]], cutlass.Int32)`.

**Why it matters.** If gemini-bot is right, either (a) this fails at first JIT compile when `FASTVIDEO_VSA_CUTEDSL=1` is exercised, or (b) it silently picks the wrong KV-block size for partial blocks — which corrupts masking only when `variable_block_sizes < kv_block_size`. The `test_vsa256_forward_vbs.py` test would catch (b) but only on Blackwell hardware with the CuTe build installed. PR body claims Blackwell verification but doesn't cite a vbs test run.

**Suggested fix.** Reply to the gemini-bot thread either: (1) "tested on Blackwell, `pytest fastvideo-kernel/tests/test_vsa256_forward_vbs.py -v` passes, here's the log" (and leave the code as-is — `kv_blk[0]` may be a benign scalar SSA shape quirk), OR (2) change to `kv_sizes[kv_blk]` if the bot is correct. Either resolution is fine; the open thread is what makes this S2.

**Evidence:** `fastvideo-kernel/python/fastvideo_kernel/block_sparse_attn_cute_fwd.py:138-141`; `fastvideo-kernel/tests/test_vsa256_forward_vbs.py`; gemini-bot PR comment, file `block_sparse_attn_cute_fwd.py` line 139, SHA `e9320ebf2b`, high-priority.

### [S2-persistent] PR description is stale relative to commits `9560e648e8` and `e9320ebf2b`

**What.** The PR body describes a `FoundationResearch/flash-attention@vsapad` submodule and a "draft for early review of dispatch design and dependency story" framing. Neither is true at HEAD:
- `.gitmodules` at HEAD has no `flash-attention` entry (removed in `9560e648e8`).
- Install path is now `pip install "git+https://github.com/Dao-AILab/flash-attention.git@c19cd20e#subdirectory=flash_attn/cute"`, documented in `fastvideo-kernel/README.md` (added in `e9320ebf2b`).
- PR is not marked draft AND label `ready` is set, contradicting the "draft for early review" body text.

**Why it matters.** Future reviewers (and automated review pipelines) read the body for context. Stale body content keeps generating already-resolved concerns ("submodule pin mutable" / "vsapad branch needed") and obscures the current architectural decision (Dao-AILab upstream pin at an immutable SHA).

**Suggested fix.** Edit the PR body to (1) drop the "draft for early review" framing or genuinely re-mark draft and remove `ready`; (2) replace the submodule sentence with the upstream-pin install snippet from `fastvideo-kernel/README.md`; (3) optionally add an explicit "Verified on Blackwell with `pytest fastvideo-kernel/tests/test_vsa256_forward*.py`" line if true.

**Evidence:** PR body vs commits `9560e648e8`, `e9320ebf2b`; `.gitmodules` at HEAD; `fastvideo-kernel/README.md:30-50`.

---

S3 findings (redundant `_force_tk` branch, redundant `FASTVIDEO_VSA_CUTEDSL` explicit check, 3-way parity tolerance justification, `torch.stack` idiom) are recorded in the local review file but not surfaced here — none are blocking.

**Strengths:** clean 64-path isolation (single `block_elements == 256` branch, no shared mutation, default `(4,4,4)` preserves the existing path byte-for-byte structurally); lazy optional-import for `flash_attn.cute` with an actionable error message naming all three required deps and the env-var; @Davids048's fork-submodule concern fully resolved in `9560e648e8` + `e9320ebf2b`; four new pytest files with `pytest.importorskip` guards on the optional CuTe build; consistent `FASTVIDEO_VSA_` env-var prefix with the legacy alias correctly mapped in exactly one place.

*— Gob (@SolitaryThinker's AI reviewer). Full review archived locally.*

**Review by SolitaryThinker — APPROVED** [review]
